pkg/workceptor/kubernetes.go

```
@@ -374,7 +374,21 @@ mainLoop:
 		for { // check between every line read to see if we need to stop reading
 			line, err := streamReader.ReadString('\n')
 			if err != nil {
-				// First check if the error is not EOF, if error is not EOF retry 5 times if error persists set error and mark the job as failed.
+				// Check if the context was canceled and the work state isn't "Succeeded".
+				// If so, set the error and mark the job as failed.
+				if kw.GetContext().Err() == context.Canceled {
+					if kw.Status().State != WorkStateSucceeded {
+						errMsg := fmt.Sprintf("Context was canceled while reading logs for pod %s/%s. This is unrecoverable. Marking the job as failed and exiting. Error: %s",
+							podNamespace,
+							podName,
+							err.Error(),
+						)
+						*stdoutErr = fmt.Errorf("%s", errMsg)
```

I remember in #1400 that we removed a check for canceled context, and that line was one of the first things @AaronH88 called out from our initial review of the log messages. I think the analysis was that we were checking the context and marking an error, but we weren't examining or logging the underlying error, so it might have masked a real error. I can't find a comment other than https://github.com/ansible/receptor/pull/1400#discussion_r2266349052.

Since this does check the context canceled and logs the error from reading the stream, I think it's covering that case. But I would like to get @AaronH88 to weigh in.

---

Yeah good points,

Their seems to be an action on us to map out all cases where the context gets canceled and improve that flow, but it would be outside the scope of this current work.

I agree with both of ye, this current check seems to be good enough for now 😄